candidategenerator.js

```diff
@@ -28,14 +28,82 @@ var wNum = 0;
    - obeysExhaustivity (boolean or array of categories at which to require conformity to exhaustivity)
    - obeysHeadedness (boolean)
    - obeysNonrecursivity (boolean)
+	 - rootCategory (string)
+	 - recursiveCategory (string)
+	 - terminalCategory (string)
    - addTones (string). Possible values include:
 	 		- "addJapaneseTones"
 			- "addIrishTones_Elfner"
 			- "addIrishTones_Kalivoda"
+	- noUnary (boolean): if true, don't create any nodes that immediately dominate only a single terminal.
+	- requireRecWrapper (boolean). Formerly "requirePhiStem"
 */
 window.GEN = function(sTree, words, options){
 	options = options || {}; // if options is undefined, set it to an empty object (so you can query its properties without crashing things)
 
+	options.recursiveCategory = options.recursiveCategory || "phi"; //sets the default of recursiveCategory option to "phi"
+
+	//set default root root category based on options passed
+	if(!options.rootCategory){ //root category is unspecifed
+		//recursiveCategory is specified as "w"
+		if(options.recursiveCategory === "w"){
+			options.rootCategory = "phi";
+		}
+		else{
+			options.rootCategory = "i";
+		}
+	}
+
+	//set default terminal category based on options passed
+	if(!options.terminalCategory){ //terminalCategory is unspecified
+		//terminalCategory unspecified, recursiveCategory specified as "i"
+		if(options.recursiveCategory === "i"){
+			options.terminalCategory = "phi";
+		}
+		//terminalCategory unspecified, recursiveCategory specified as "w"
+		else if(options.recursiveCategory === "w"){
+			options.terminalCategory = "Ft";
+		}
+		//neither terminalCategory nor recursiveCategory is specified, default to "w"
+		else{
+			options.terminalCategory = "w";
+		}
+	}
+
+	/* the prosodic hierarchy should include the categories specified in
+	 * options.rootCategory, options.recursiveCategory and options.terminalCategory
+	 */
+	if(options.rootCategory || options.recursiveCategory || options.terminalCategory){
+		//the specified root category should probably be highest
+		if(pCat.indexOf(options.rootCategory)<0)
+			pCat.unshift(options.rootCategory);
+		if(pCat.indexOf(options.recursiveCategory)<0 && options.rootCategory !== options.recursiveCategory){
+			//the specified recursive category should be after iota
```

Comment:
Again, there should be a warning that the requested recursive category wasn't in pCat.
I'm not really so sure that GEN should be able to modify pCat like this anyway...

Reply:
I think that if the user enters unrecognized categories, we should actually just throw an error and not give any output, because it's impossible to know where in the prosodic hierarchy they should really be added. I see that you've made some logical choices about where in the hierarchy they could be added but they still seem somewhat arbitrary to  me, and will certainly yield a very opaque prosodic hierarchy! Let's pull out this code and stash it somewhere. It could be good to add functionality to let users create custom prosodic hierarchies and *then* be able to use them in their GEN.

Reply:
As mentioned above, GEN can no longer modify pCat, a warning pops up and an error is thrown instead. These changes will be pushed later today

Reply:
Sounds good.